src/languages/java.js

```diff
@@ -18,32 +20,53 @@ export default function(hljs) {
   var ANNOTATION = {
     className: 'meta',
     begin: '@' + JAVA_IDENT_RE,
-    contains:[
+    contains: [
       {
         begin: /\(/,
         end: /\)/,
         contains: ["self"] // allow nested () inside our annotation
       },
     ]
-  }
-  // https://docs.oracle.com/javase/7/docs/technotes/guides/language/underscores-literals.html
-  var JAVA_NUMBER_RE = '\\b' +
-    '(' +
-      '0[bB]([01]+[01_]+[01]+|[01]+)' + // 0b...
-      '|' +
-      '0[xX]([a-fA-F0-9]+[a-fA-F0-9_]+[a-fA-F0-9]+|[a-fA-F0-9]+)' + // 0x...
-      '|' +
-      '(' +
-        '([\\d]+[\\d_]+[\\d]+|[\\d]+)(\\.([\\d]+[\\d_]+[\\d]+|[\\d]+))?' +
-        '|' +
-        '\\.([\\d]+[\\d_]+[\\d]+|[\\d]+)' +
-      ')' +
-      '([eE][-+]?\\d+)?' + // octal, decimal, float
-    ')' +
-    '[lLfF]?';
+  };
+  /**
+   * A given sequence, possibly with underscores
+   * @type {(s: string | RegExp) => string}  */
+  var SEQUENCE_ALLOWING_UNDERSCORES = (seq) => regex.concat('[', seq, ']+([', seq, '_]*[', seq, ']+)?');
   var JAVA_NUMBER_MODE = {
     className: 'number',
-    begin: JAVA_NUMBER_RE,
+    variants: [
+      { begin: `\\b(0[bB]${SEQUENCE_ALLOWING_UNDERSCORES('01')})[lL]?` }, // binary
+      { begin: `\\b(0${SEQUENCE_ALLOWING_UNDERSCORES('0-7')})[dDfFlL]?` }, // octal
+      {
+        begin: regex.concat(
+          /\b0[xX]/,
+          regex.either(
+            regex.concat(SEQUENCE_ALLOWING_UNDERSCORES('a-fA-F0-9'), /\./, SEQUENCE_ALLOWING_UNDERSCORES('a-fA-F0-9')),
+            regex.concat(SEQUENCE_ALLOWING_UNDERSCORES('a-fA-F0-9'), /\.?/),
+            regex.concat(/\./, SEQUENCE_ALLOWING_UNDERSCORES('a-fA-F0-9')),
+          ),
+          /([pP][+-]?(\d+))?/,
+          /[fFdDlL]?/ // decimal & fp mixed for simplicity
+        )
+      },
+      // scientific notation
+      { begin: regex.concat(
+        /\b/,
+        regex.either(
+          /\d*\.[\d_]*\d+/, // 3.3e2
+          /\d+[\d_]*/ // 3e2
```

Comment:
On these two lines `e2` in the comment doesn't look right, 'cause the regexps are for the numeric part only.
Also I'd change them to use `SEQUENCE_ALLOWING_UNDERSCORES(/\d/)` to be consistent with everything else.
Btw, is it correct that the first regex doesn't allow underscores before the dot?..

Reply:
I'm a little bit past caring at this point. :-)  I'll take a quick look but I'm not sure these are the same pattern as SEQUENCE_ALLOWING_UNDERSCORES (or should be).

I'll take a look at the comments also and see if anything occurs to me on improving them.

Reply:
You were right, they could be simplified.